<a href="https://colab.research.google.com/github/LC1332/Speaker-Grouping/blob/main/notebook/%E6%89%B9%E9%87%8F%E6%8A%BD%E5%8F%96%E5%9B%BE%E7%89%87%E5%92%8C%E9%9F%B3%E9%A2%91%E7%89%B9%E5%BE%81.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- [ ] 载入并解压数据
- [ ] 批量抽取clip中的vit特征
- [ ] 批量抽取音频特征
- [ ] 保存特征为parquet文件

# 载入数据并解压

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!ls /content/drive/MyDrive/Speaker/data

liangjian_10.zip


In [3]:
zip_name = "/content/drive/MyDrive/Speaker/data/liangjian_10.zip"

我希望将 zip_name = "/content/drive/MyDrive/Speaker/data/liangjian_10.zip"

这个文件copy到/content/下 并且解压到/content/{zip_file_name}

要如何用python实现

In [4]:
import shutil
import zipfile
import os

# 定义源 ZIP 文件路径和目标路径
zip_path = "/content/drive/MyDrive/Speaker/data/liangjian_10.zip"
target_dir = "/content/"

# 复制 ZIP 文件到指定目录
shutil.copy(zip_path, target_dir)

# 构建解压后的目标文件夹路径
zip_file_name = os.path.basename(zip_path)
unzip_folder = os.path.join(target_dir, zip_file_name.replace('.zip', ''))

# 创建目标文件夹
if not os.path.exists(unzip_folder):
    os.makedirs(unzip_folder)

# 解压 ZIP 文件
with zipfile.ZipFile(os.path.join(target_dir, zip_file_name), 'r') as zip_ref:
    zip_ref.extractall(unzip_folder)

print(f'文件已解压到 {unzip_folder}')


文件已解压到 /content/liangjian_10


In [5]:
!ls /content/liangjian_10/亮剑10

audios	meta.jsonl  screeshots


In [6]:
source_dir = "/content/liangjian_10/亮剑10"

# 解析meta

In [7]:
import pandas as pd

# 定义存有 JSONL 文件的目录
source_dir = "/content/liangjian_10/亮剑10"
jsonl_file = f"{source_dir}/meta.jsonl"

# 读取 JSONL 文件到 pandas DataFrame
# lines=True 表示文件中的每一行都是一个独立的 JSON 对象
meta_data = pd.read_json(jsonl_file, lines=True)

# 显示前 5 行数据
print(meta_data.head())


   人物      人物台词          开始时间          结束时间                 audio_file  \
0  段鹏  八路打鬼子不含糊  00:01:34.140  00:01:35.340  outdata/亮剑10/audios/0.wav   
1  段鹏      可俺不行  00:01:35.340  00:01:37.060  outdata/亮剑10/audios/1.wav   
2  段鹏    俺家里有老娘  00:01:37.060  00:01:38.340  outdata/亮剑10/audios/2.wav   
3  段鹏       俺走了  00:01:38.340  00:01:39.320  outdata/亮剑10/audios/3.wav   
4  段鹏      俺娘咋办  00:01:39.320  00:01:40.100  outdata/亮剑10/audios/4.wav   

                  screeshot_file  
0  outdata/亮剑10/screeshots/0.jpg  
1  outdata/亮剑10/screeshots/1.jpg  
2  outdata/亮剑10/screeshots/2.jpg  
3  outdata/亮剑10/screeshots/3.jpg  
4  outdata/亮剑10/screeshots/4.jpg  


# 抽取图片特征

In [ ]:
def clean_name( file_name ):
    return file_name.replace("outdata/亮剑10/", "")

In [9]:
!wget -q https://raw.githubusercontent.com/LC1332/simple-face-recognition/main/CLIPExtractor.py

In [12]:
screenshots = meta_data["screeshot_file"].values
screenshots = [clean_name(f) for f in screenshots]
screenshots = [os.path.join(source_dir, f) for f in screenshots]

print(screenshots[:5])

['/content/liangjian_10/亮剑10/screeshots/0.jpg', '/content/liangjian_10/亮剑10/screeshots/1.jpg', '/content/liangjian_10/亮剑10/screeshots/2.jpg', '/content/liangjian_10/亮剑10/screeshots/3.jpg', '/content/liangjian_10/亮剑10/screeshots/4.jpg']


In [13]:
from CLIPExtractor import CLIPExtractor

visual_extractor = CLIPExtractor()

config.json:   0%|          | 0.00/4.10k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/599M [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/905 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/961k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/525k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.22M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/389 [00:00<?, ?B/s]

In [14]:
visual_feature = visual_extractor.extract(screenshots)

100%|██████████| 47/47 [00:17<00:00,  2.66it/s]


---

# 抽取音频特征

In [20]:
!pwd

/content


In [21]:

!git clone --filter=blob:none --no-checkout https://github.com/LC1332/Legacy-Haruhi-1.0
%cd Legacy-Haruhi-1.0
!git sparse-checkout init --cone
!git sparse-checkout set yuki_builder/audio_feature_ext
!git checkout


fatal: destination path 'Legacy-Haruhi-1.0' already exists and is not an empty directory.
/content/Legacy-Haruhi-1.0
remote: Enumerating objects: 25, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 25 (delta 0), reused 1 (delta 0), pack-reused 24
Receiving objects: 100% (25/25), 50.88 KiB | 854.00 KiB/s, done.
Your branch is up to date with 'origin/main'.


In [22]:
%cd /content/

/content


In [29]:
%cd /content/Legacy-Haruhi-1.0/yuki_builder

/content/Legacy-Haruhi-1.0/yuki_builder


!ls Legacy-Haruhi-1.0/yuki_builder/audio_feature_ext

下的内容有audio_fea_ext.py  data_utils  incredata.py  __init__.py  modules  tool.py  utils

我想载入audio_fea_ext.py中对应的AudioFeatureExtract类，我要怎么做

In [34]:
# 现在可以导入模块和类
from audio_feature_ext.audio_fea_ext import AudioFeatureExtraction

# 创建类的实例
afe = AudioFeatureExtraction()


downloading model pth model.pth
model.pth success download
downloading model pth model.state
model.state success download
downloading model pth optimizer.pth
optimizer.pth success download
成功加载模型参数和优化方法参数：./audio_feature_ext/models/model.pth


# 现在可以导入模块和类
from audio_feature_ext.audio_fea_ext import AudioFeatureExtraction

# 创建类的实例
afe = AudioFeatureExtraction()

这个程序可以正常运行，并且这个类中有一个函数

    def extract_features(self, root_dir):
        sub_dirs = get_subdir(root_dir)

        for dir in sub_dirs[:]:
            voice_files = get_filename(os.path.join(dir, 'voice'))
            for file, pth in voice_files:
                new_dir = os.path.join(dir, 'feature')
                os.makedirs(new_dir, exist_ok=True)
                try:
                    feature = self.infer(pth)[0]
                    with open(f"{new_dir}/{file}.pkl", "wb") as f:
                        pickle.dump(feature, f)
                except:
                    continue
        print('音频特征提取完成')

我现在希望继承AudioFeatureExtraction这个类，产生一个新类NewAudioFeatureExtraction

这个类存在一个函数extract_from_files(self, voice_files)

并且直接将features以list形式进行return，不需要保存到文件，请为我实现继承类部分的代码

In [42]:
from audio_feature_ext.audio_fea_ext import AudioFeatureExtraction
from tqdm import tqdm

class NewAudioFeatureExtraction(AudioFeatureExtraction):
    def extract_from_files(self, voice_files):
        features = []
        for file in tqdm(voice_files):
            try:
                feature = self.infer(file)[0]
                features.append(feature)
            except Exception as e:
                print(f"Error processing file {file}: {str(e)}")
                features.append([])
                continue
        return features

voice_extractor = NewAudioFeatureExtraction()

成功加载模型参数和优化方法参数：./audio_feature_ext/models/model.pth


In [43]:
audio_files = meta_data["audio_file"].values
audio_files = [clean_name(f) for f in audio_files]
audio_files = [os.path.join(source_dir, f) for f in audio_files]

print(audio_files[:5])

['/content/liangjian_10/亮剑10/audios/0.wav', '/content/liangjian_10/亮剑10/audios/1.wav', '/content/liangjian_10/亮剑10/audios/2.wav', '/content/liangjian_10/亮剑10/audios/3.wav', '/content/liangjian_10/亮剑10/audios/4.wav']


In [45]:
voice_features = voice_extractor.extract_from_files( audio_files)

100%|██████████| 752/752 [00:21<00:00, 35.29it/s]


In [47]:
print(len(voice_features))

752


---

# 保存为parquet文件

# 显示前 5 行数据
print(meta_data.head())

的结果如下

   人物      人物台词          开始时间          结束时间                 audio_file  \
0  段鹏  八路打鬼子不含糊  00:01:34.140  00:01:35.340  outdata/亮剑10/audios/0.wav   
1  段鹏      可俺不行  00:01:35.340  00:01:37.060  outdata/亮剑10/audios/1.wav   
2  段鹏    俺家里有老娘  00:01:37.060  00:01:38.340  outdata/亮剑10/audios/2.wav   
3  段鹏       俺走了  00:01:38.340  00:01:39.320  outdata/亮剑10/audios/3.wav   
4  段鹏      俺娘咋办  00:01:39.320  00:01:40.100  outdata/亮剑10/audios/4.wav   

                  screeshot_file  
0  outdata/亮剑10/screeshots/0.jpg  
1  outdata/亮剑10/screeshots/1.jpg  
2  outdata/亮剑10/screeshots/2.jpg  
3  outdata/亮剑10/screeshots/3.jpg  
4  outdata/亮剑10/screeshots/4.jpg  

我已经将screenshot转变为了visual_feature

visual_feature是一个和meta_data长度一样的list of list of float

同时已经将audio_file转变为了voice_features

voice_features是一个和meta_data长度一样的list of list of float

我希望实现一段python程序，将meta_data的每一行增加visual_feature 和audio_feature字段，并把对应的特征存储进去

最后，将整个文件存储在save_name = "/content/liangjian_10_feature.parquet"文件中



In [48]:
import pandas as pd

# 假设 meta_data, visual_feature, voice_features 已经定义好了

# 转换 list of list 到 pandas Series
visual_feature_series = pd.Series(visual_feature)
voice_feature_series = pd.Series(voice_features)

# 将特征添加为新列
meta_data['visual_feature'] = visual_feature_series
meta_data['audio_feature'] = voice_feature_series

# 指定保存的文件名
save_name = "/content/liangjian_10_feature.parquet"

# 保存 DataFrame 为 Parquet 文件
meta_data.to_parquet(save_name)

print("特征已成功保存到文件中")


特征已成功保存到文件中
